File: antenna-azimuth-mapper/azimuth_mapper/solar.py
"""Low-precision solar position (Michalsky/NOAA-style), no external dependencies.

Accurate to roughly 0.01 deg for 1950-2050, which is far more than this tool
needs: it is only used to sanity-check a shadow direction detected in an
image, not for anything safety-critical.
"""

from __future__ import annotations

import math
from dataclasses import dataclass
from datetime import datetime, timezone


@dataclass(frozen=True)
class SolarPosition:
    azimuth_deg: float  # compass bearing of the sun, 0 = north, clockwise
    elevation_deg: float  # angle above the horizon


def _julian_day(dt_utc: datetime) -> float:
    if dt_utc.tzinfo is None:
        dt_utc = dt_utc.replace(tzinfo=timezone.utc)
    dt_utc = dt_utc.astimezone(timezone.utc)
    epoch = datetime(2000, 1, 1, 12, 0, 0, tzinfo=timezone.utc)  # JD 2451545.0
    return 2451545.0 + (dt_utc - epoch).total_seconds() / 86400.0


def solar_position(lat_deg: float, lon_deg: float, dt_utc: datetime) -> SolarPosition:
    """Sun azimuth/elevation as seen from (lat, lon) at the given UTC instant."""
    jd = _julian_day(dt_utc)
    n = jd - 2451545.0

    mean_lon = math.radians((280.460 + 0.9856474 * n) % 360)
    mean_anomaly = math.radians((357.528 + 0.9856003 * n) % 360)
    ecliptic_lon = mean_lon + math.radians(
        1.915 * math.sin(mean_anomaly) + 0.020 * math.sin(2 * mean_anomaly)
    )
    obliquity = math.radians(23.439 - 0.0000004 * n)

    right_ascension = math.atan2(
        math.cos(obliquity) * math.sin(ecliptic_lon), math.cos(ecliptic_lon)
    )
    declination = math.asin(math.sin(obliquity) * math.sin(ecliptic_lon))

    gmst_deg = (280.46061837 + 360.98564736629 * n) % 360
    hour_angle_deg = (gmst_deg + lon_deg - math.degrees(right_ascension)) % 360
    if hour_angle_deg > 180:
        hour_angle_deg -= 360
    hour_angle = math.radians(hour_angle_deg)

    lat = math.radians(lat_deg)
    sin_elevation = math.sin(lat) * math.sin(declination) + math.cos(lat) * math.cos(
        declination
    ) * math.cos(hour_angle)
    sin_elevation = max(-1.0, min(1.0, sin_elevation))
    elevation = math.asin(sin_elevation)

    cos_azimuth = (math.sin(declination) - math.sin(lat) * sin_elevation) / (
        math.cos(lat) * math.cos(elevation)
    ) if abs(math.cos(lat) * math.cos(elevation)) > 1e-9 else 0.0
    cos_azimuth = max(-1.0, min(1.0, cos_azimuth))
    azimuth = math.acos(cos_azimuth)
    azimuth_deg = math.degrees(azimuth)
    if math.sin(hour_angle) > 0:
        azimuth_deg = 360 - azimuth_deg

    return SolarPosition(azimuth_deg=azimuth_deg % 360, elevation_deg=math.degrees(elevation))

File: antenna-azimuth-webapp/lib/solar.ts
/**
 * Low-precision solar position (Michalsky/NOAA-style), no external
 * dependencies. Ported 1:1 from
 * antenna-azimuth-mapper/azimuth_mapper/solar.py.
 *
 * Accurate to roughly 0.01 deg for 1950-2050 — far more than this tool
 * needs: it only sanity-checks a shadow direction detected in an image,
 * not anything safety-critical.
 */

export interface SolarPosition {
  azimuthDeg: number; // compass bearing of the sun, 0 = north, clockwise
  elevationDeg: number; // angle above the horizon
}

function toRad(deg: number): number {
  return (deg * Math.PI) / 180;
}

function toDeg(rad: number): number {
  return (rad * 180) / Math.PI;
}

function mod360(deg: number): number {
  return ((deg % 360) + 360) % 360;
}

const JD_2000_EPOCH_MS = Date.UTC(2000, 0, 1, 12, 0, 0); // JD 2451545.0

function julianDay(dtUtc: Date): number {
  return 2451545.0 + (dtUtc.getTime() - JD_2000_EPOCH_MS) / 86_400_000;
}

/** Sun azimuth/elevation as seen from (lat, lon) at the given UTC instant. */
export function solarPosition(latDeg: number, lonDeg: number, dtUtc: Date): SolarPosition {
  const jd = julianDay(dtUtc);
  const n = jd - 2451545.0;

  const meanLon = toRad(mod360(280.46 + 0.9856474 * n));
  const meanAnomaly = toRad(mod360(357.528 + 0.9856003 * n));
  const eclipticLon =
    meanLon + toRad(1.915 * Math.sin(meanAnomaly) + 0.02 * Math.sin(2 * meanAnomaly));
  const obliquity = toRad(23.439 - 0.0000004 * n);

  const rightAscension = Math.atan2(
    Math.cos(obliquity) * Math.sin(eclipticLon),
    Math.cos(eclipticLon)
  );
  const declination = Math.asin(Math.sin(obliquity) * Math.sin(eclipticLon));

  const gmstDeg = mod360(280.46061837 + 360.98564736629 * n);
  let hourAngleDeg = mod360(gmstDeg + lonDeg - toDeg(rightAscension));
  if (hourAngleDeg > 180) hourAngleDeg -= 360;
  const hourAngle = toRad(hourAngleDeg);

  const lat = toRad(latDeg);
  let sinElevation =
    Math.sin(lat) * Math.sin(declination) +
    Math.cos(lat) * Math.cos(declination) * Math.cos(hourAngle);
  sinElevation = Math.max(-1.0, Math.min(1.0, sinElevation));
  const elevation = Math.asin(sinElevation);

  const denom = Math.cos(lat) * Math.cos(elevation);
  let cosAzimuth =
    Math.abs(denom) > 1e-9
      ? (Math.sin(declination) - Math.sin(lat) * sinElevation) / denom
      : 0.0;
  cosAzimuth = Math.max(-1.0, Math.min(1.0, cosAzimuth));
  let azimuthDeg = toDeg(Math.acos(cosAzimuth));
  if (Math.sin(hourAngle) > 0) azimuthDeg = 360 - azimuthDeg;

  return { azimuthDeg: mod360(azimuthDeg), elevationDeg: toDeg(elevation) };
}
